lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,421 +25,546 @@
+        entry.emplace_back(sub.getValue(), sub.getFieldID() + fieldID);
+      // Update the map at the end, to avoid invaliding the iterator.
+      valToFieldRefs[result].append(entry.begin(), entry.end());
+      return;
+    }
+    // Break cycles from registers.
+    if (auto *def = base.getDefiningOp()) {
+      if (isa<RegOp, RegResetOp, SubfieldOp, SubindexOp, SubaccessOp>(def))
         return;
-      auto memPort = ref.getValue();
-      auto type = type_cast<BundleType>(memPort.getType());
-      auto enableFieldId = type.getFieldID((unsigned)ReadPortSubfield::en);
-      auto dataFieldId = type.getFieldID((unsigned)ReadPortSubfield::data);
-      auto addressFieldId = type.getFieldID((unsigned)ReadPortSubfield::addr);
-      if (ref.getFieldID() == enableFieldId ||
-          ref.getFieldID() == addressFieldId) {
-        for (auto dataField : fieldToVals[FieldRef(memPort, dataFieldId)])
-          children.push_back(dataField);
-      }
     }
+    valToFieldRefs[result].emplace_back(base, fieldID);
   }
 
-  void reportLoopFound(Value childVal, VisitingSet visiting) {
-    // TODO: Work harder to provide best information possible to user,
-    // especially across instances or when we trace through aliasing values.
-    // We're about to exit, and can afford to do some slower work here.
-    auto getName = [&](Value v) {
-      if (isa_and_nonnull<SubfieldOp, SubindexOp, SubaccessOp>(
-              v.getDefiningOp())) {
-        assert(!valRefersTo[v].empty());
-        // Pick representative of the "alias set", not deterministic.
-        return getFieldName(*valRefersTo[v].begin()).first;
+  void handleMemory(MemOp mem) {
+    if (mem.getReadLatency() > 0)
+      return;
+    // Add the enable and address fields as the reaching definitions to the data
+    // field.
+    for (auto memPort : mem.getResults())
+      // TODO: Can reftype ports create cycle ?
+      if (auto type = type_dyn_cast<BundleType>(memPort.getType())) {
+        auto enableFieldId = type.getFieldID((unsigned)ReadPortSubfield::en);
+        auto addressFieldId = type.getFieldID((unsigned)ReadPortSubfield::addr);
+        auto dataFieldId = type.getFieldID((unsigned)ReadPortSubfield::data);
+        addReachingDef({memPort, static_cast<unsigned int>(dataFieldId)},
+                       {memPort, static_cast<unsigned int>(enableFieldId)});
+        addReachingDef({memPort, static_cast<unsigned int>(dataFieldId)},
+                       {memPort, static_cast<unsigned int>(addressFieldId)});
       }
-      return getFieldName(FieldRef(v, 0)).first;
+  }
+
+  // Perform an iterative DFS traversal of the given graph. Record paths between
+  // the ports and detect and report any cycles in the graph.
+  LogicalResult dfsTraverse(const ReachingDefgraphType &graph) {
+    auto numNodes = graph.size();
+    SmallVector<bool> onStack(numNodes, false);
+    SmallVector<unsigned> dfsStack;
+
+    auto hasCycle = [&](unsigned rootNode, DenseSet<unsigned> &visited,
```

Comment:
Nit: Why is `visited` separate from `onStack` and `dfsStack` re:what's captured and what's passed into this function?

Ah, `onStack`/`dfsStack` are empty between invocations.  That also answers question why it's okay to clear `visited` without clearing `onStack`.

Reply:
Right, this is required for the iterative implementation to avoid the dfs recursion.